### PR TITLE
test(modules): implement first tests of this repo

### PR DESCRIPTION
**Pull request by leiicamundi**
# Description

This pull request introduces testing for the modules, covering the following scenarios:
- Standard EKS values
- Upgrade of an EKS cluster
- Integration of EKS and RDS with custom values

The tests are organized into dedicated test suites and can be run in parallel using a separate copy of the module folder. By default, resources are cleaned up at the end of the tests.

Automation:
- A nightly cleanup of cloud resources is scheduled for 5 AM every day.
- Tests are triggered in the CI pipeline upon pull request approval and on-demand. Additionally, they run every Tuesday at 2 AM.

Please note that tests may take a considerable amount of time and may occasionally fail depending on available AWS resources.

~~**Note**: To initiate tests, the pull request must be labeled with "testing-allowed". Otherwise, tests will not be launched and will be marked as failed. Either "testing-allowed" or "testing-ci-not-necessary" labels must be added, the last one allows to skip tests.~~

## Known Limitations

- Due to a maximum policy on Elastic IPs, tests cannot be run in parallel. (not for multiple workflows but works for a single workflow with multiple tests in //)
- IRSA integration is not included in this PR. Proper setup of ServiceAccounts, etc., would be required. If deemed relevant, we can add it in subsequent iterations.

Also implement https://github.com/camunda/team-infrastructure-experience/issues/160

### CODE REVIEW

File: .github/workflows/lint.yml
Comment (Langleu) [inline .github/workflows/lint.yml:24]: maybe makes sense to switch to using asdf for that?
https://github.com/asdf-vm/actions?tab=readme-ov-file#setup

Generally Python is available on the runner, not sure why to specifically add it but nevertheless, could be added to `.tool-versions` as well.

Results in you not having to maintain the versions in different files.

File: .github/workflows/nightly_cleanup.yml
Comment (Langleu) [inline .github/workflows/nightly_cleanup.yml:60]: I'd go with a slightly higher number like 6 hours here and line 78.
It's not guaranteed that your GitHub Action Job is scheduled on point and 4 hours is the exact time difference between the two jobs, so might run into an unforeseen edge case that it may not clean up everything.

File: .github/workflows/tests.yml
Comment (Langleu) [inline .github/workflows/tests.yml:31]: here and following lines.
I don't understand the purpose of `testing-ci-not-necessary`.

File: .github/workflows/tests.yml
Comment (Langleu) [inline .github/workflows/tests.yml:53]: if you only need those two specific tools.
I'd recommend getting the version from `.tool-verisons.`

https://github.com/endorama/asdf-parse-tool-versions

File: .github/workflows/lint.yml
Comment (Langleu) [reply .github/workflows/lint.yml:24]: or as mentioned in another comment, pull the versions from `.tool-versions` https://github.com/endorama/asdf-parse-tool-versions

File: test/.gitignore
Comment (Langleu) [inline test/.gitignore:1]: why not include it in the global `.gitignore` in root?

File: test/src/.tool-versions
Comment (Langleu) [inline test/src/.tool-versions:1]: I'd suggest moving it to the global `.tool-versions` in the root.
It's easier for users if there's a single source of truth.

File: .github/workflows/tests.yml
Comment (maxdanilov) [reply .github/workflows/tests.yml:11]: imo, this will make this workflow easier to read, as there's be only one `if` condition in the beginning

**Review by maxdanilov — CHANGES_REQUESTED** [review]
Overall I like the approach for the test organization and the basic scenarios to cover with Terratest 👏 
Leaving some comments to clarify my understanding and make some suggestions.

Also, I've run the tests locally and somehow it showed be a failure though everything looks fine in the first glance:

```
TestUpgradeEKSTestSuite/TestUpgradeEKS 2024-04-11T19:34:24+02:00 logger.go:66:
TestUpgradeEKSTestSuite/TestUpgradeEKS 2024-04-11T19:34:24+02:00 logger.go:66: No changes. No objects need to be destroyed.
TestUpgradeEKSTestSuite/TestUpgradeEKS 2024-04-11T19:34:24+02:00 logger.go:66:
TestUpgradeEKSTestSuite/TestUpgradeEKS 2024-04-11T19:34:24+02:00 logger.go:66: Either you have not created any objects yet or the existing objects were
TestUpgradeEKSTestSuite/TestUpgradeEKS 2024-04-11T19:34:24+02:00 logger.go:66: already deleted outside of Terraform.
TestUpgradeEKSTestSuite/TestUpgradeEKS 2024-04-11T19:34:25+02:00 logger.go:66:
TestUpgradeEKSTestSuite/TestUpgradeEKS 2024-04-11T19:34:25+02:00 logger.go:66: Destroy complete! Resources: 0 destroyed.
TestUpgradeEKSTestSuite/TestUpgradeEKS 2024-04-11T19:34:25+02:00 logger.go:66:
    upgrade_eks_test.go:62: Cleaning up resources...
--- PASS: TestUpgradeEKSTestSuite (3553.54s)
    --- PASS: TestUpgradeEKSTestSuite/TestUpgradeEKS (3553.54s)
FAIL
exit status 1
FAIL	github.com/camunda/camunda-tf-eks-module	3555.519s
make: *** [test] Error 1
```
the same test is failing in the [CI](https://github.com/camunda/camunda-tf-eks-module/actions/runs/8646047218/job/23704492156?pr=30):

```
TestCustomEKSRDSTestSuite/TestCustomEKSAndRDS 2024-04-11T12:13:38Z logger.go:66: aws_rds_cluster_instance.aurora_instance[0]: Still creating... [6m50s elapsed]
Update status: Successful
Update completed successfully

=== FAIL: . TestDefaultEKSTestSuite (1824.86s)

DONE 6 tests, 2 failures in 3705.742s
Error: Process completed with exit code 1.
```

----
Overall I'm finding it hard to see what exactly failed (what assertion etc.) - is there a way to make Terratest output log more clear in that regards?

File: .github/workflows/tests.yml
Comment (Langleu) [reply .github/workflows/tests.yml:6]: it's about a scheduled job. Renovate isn't impacted by that since `pull_request` will run regardless.
Nevertheless, can't hurt to put it to Monday as sanity check that whatever was auto-merged is working all-together.

File: .github/workflows/tests.yml
Comment (Langleu) [reply .github/workflows/tests.yml:32]: sounds to me like that.
Even if it's auto labeled via a different workflow, it will likely be too slow for this workflow to catch it.

File: .github/workflows/tests.yml
Comment (Langleu) [reply .github/workflows/tests.yml:54]: yes, same works for go.

File: .github/workflows/tests.yml
Comment (Langleu) [reply .github/workflows/tests.yml:32]: I still don't understand why we need it in the first place.

File: .github/workflows/tests.yml
Comment (Langleu) [reply .github/workflows/tests.yml:89]: if you want to use `gotestsum`, I'd suggest going with the asdf approach.
- add it to the `.tool-versions` - (for local usage - https://github.com/araspro/asdf-gotestsum)
- pull the version from `.tool-versions`
- https://github.com/marketplace/actions/gotestsum-installer

File: test/README.md
Comment (Langleu) [inline test/README.md:5]: I'd change this part to just terraform.
I wouldn't impose a strict requirement on either one. We generally test with terraform and if they diverge in the future this may be misleading.

File: test/README.md
Comment (Langleu) [inline test/README.md:1]: I'd propose to generalise it, how people set it up is up to them.

Just list the requirements
- golang
- terraform
- aws-cli
- apparently eksctl?

File: test/README.md
Comment (Langleu) [reply test/README.md:5]: additionally in the docs we generally speak about `terraform` and `opentofu` may be confusing.
I'd stay with `terraform` for now to be consistent on a company level.
If you use `opentofu` locally and it works, no problem.

Comment (Langleu) [reply test/README.md:5]: I'm aware of the license change but I see it's very ambiguous on how it can be laid out.
Would you suggest changing the docs to opentofu as well?

I guess it's a general question we have to solve within the company whether we'll focus on Terraform or OpenTofu as they may diverge at some point. Same as we've observed with Elastic and OpenSearch.

**Comment by Langleu** [comment]
> Due to a maximum policy on Elastic IPs, tests cannot be run in parallel.

What speaks against increasing the quota by a reasonable number?
I did the same in `eu-west-2` and `eu-west-3`

I was thinking about KMS a bit, how about just simply trying to schedule deletion for everything and accepting errors  / ignoring errors `|| true`. It will schedule deletion for the keys that are obsolete and has access to and the rest is simply ignored but ultimately ensuring that whatever the user created is removed.

File: test/src/upgrade_eks_test.go
Comment (Langleu) [inline test/src/upgrade_eks_test.go:145]: since you don't use the localPort1 anywhere else nor rely on it in any way, you can also consider just setting it in the parameter as `0` since then terratest will automatically determine an available port on the host system.

File: test/src/upgrade_eks_test.go
Comment (Langleu) [inline test/src/upgrade_eks_test.go:208]: same mentioned as above on setting it simply to `0` to leave the task to terratest on finding an available port.

**Comment by Langleu** [comment]
I'm still a bit confused on the whole labeling topic and whether we need it. Might make sense to sync on that topic in a quick call.
E.g. trigger GHA by file changes --> https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#running-your-pull_request-workflow-based-on-files-changed-in-a-pull-request
Outside contributors would anyway need an approval.

**Review by Langleu — CHANGES_REQUESTED** [review]
Overall very good job!
Mostly things about GHA or wordings here and there.
I really like your test approach, quite interesting to see!

File: .github/workflows/tests.yml
Comment (Langleu) [reply .github/workflows/tests.yml]: if you want to keep it, then I'd go with a solution like here: https://github.com/camunda/web-modeler/blob/main/.github/workflows/preview-env-teardown.yml#L10
It's a bit cleaner on a job level than every single step.

Also probably a single label is more than sufficient to decide on whether to trigger or not.
Additionally, one could add an exception for Renovate (`GITHUB_ACTOR`)

Comment (maxdanilov) [reply .github/workflows/tests.yml]: second thought on the opinion, in previous companies we just had a label called `skip-ci` that was a universal way to claim should be skipped in the test - maybe we can just use this one label here, and do the filtering on the job level (e.g. if the label `skip-ci` is present -> don't run the terratest checks)

Comment (Langleu) [reply .github/workflows/tests.yml]: or just simply rely on the built-in capabilities --> https://docs.github.com/en/actions/managing-workflow-runs/skipping-workflow-runs

File: .github/workflows/tests.yml
Comment (maxdanilov) [reply .github/workflows/tests.yml:2]: not sure about dynamic population, afraid that it wound overcomplicate things instead of just having 2 separate workflows (but also not sure how much duplication it would be in that case)

File: test/README.md
Comment (maxdanilov) [reply test/README.md:5]: > Using Terraform with the new BSL license prevents us from embedding it in the product (source: [DuploCloud](https://duplocloud.com/blog/terraform-license-change-impacts-devops/)).

I think it would not be a problem, as we're not redistributing Terraform itself neither make profits of it by building a product around it.

Comment (maxdanilov) [reply test/README.md:5]: > I guess it's a general question we have to solve within the company whether we'll focus on Terraform or OpenTofu as they may diverge at some point. Same as we've observed with Elastic and OpenSearch.

Agree, if it's not a big effort, I'd suggest using usual Terraform for now.

File: modules/fixtures/postgres-client.bak.yml
Comment (maxdanilov) [reply modules/fixtures/postgres-client.bak.yml]: I think it's fine to skip the IRSA tests for now, we can always add them later.

File: .github/workflows/lint.yml
Comment (Langleu) [inline .github/workflows/lint.yml:16]: still not sure why you use `setup-python`, as the output says:
```
Warning: Neither 'python-version' nor 'python-version-file' inputs were supplied. Attempting to find '.python-version' file.
Warning: .python-version doesn't exist.
Warning: The `python-version` input is not set.  The version of Python currently in `PATH` will be used.
```

File: .github/workflows/lint.yml
Comment (Langleu) [inline .github/workflows/lint.yml:22]: I'd still prefer to pull this version from the `.tool-versions` via e.g. https://github.com/endorama/asdf-parse-tool-versions

File: .github/workflows/lint.yml
Comment (Langleu) [inline .github/workflows/lint.yml:23]: same as above, would suggest to supply the version from `.tool-versions`
```
Warning: go-version input was not specified. The action will try to use pre-installed version.
go version go1.21.9 linux/amd64
```

even though you're using 1.22

File: .github/workflows/lint.yml
Comment (Langleu) [reply .github/workflows/lint.yml:24]: It's still the same.

I'd suggest pulling the version from `.tool-versions` and then installing like

```
    - name: Install Terraform Docs
      uses: jaxxstorm/action-install-gh-release@v1.10.0
      with:
        repo: terraform-docs/terraform-docs
        tag: v0.17.0
```

File: .github/workflows/nightly_cleanup.yml
Comment (Langleu) [reply .github/workflows/nightly_cleanup.yml:12]: It's fixed in the `tests.yml` but not this file `nightly_cleaup.yml`

File: .github/workflows/tests.yml
Comment (Langleu) [inline .github/workflows/tests.yml:146]: just a suggestion to define retention days, not sure we need them by default for 90 days?

File: .github/workflows/tests.yml
Comment (Langleu) [reply .github/workflows/tests.yml:19]: hehe, the `developer.md` part happened in just that second ^^, but still the comment here could probably be aligned to mention commit message

File: .github/workflows/tests.yml
Comment (Langleu) [reply .github/workflows/tests.yml:6]: I'm fine keeping that on Tuesday.
Assumption: Renovate runs on the weekend. On Monday we merge the PRs / check / adjust depending on wh at Renovate says.
Tuesday kicks-in as sanity check to whatever was done on Monday.

Maybe we figure out later that we want to run it more often, we'll see.

**Review by Langleu — APPROVED** [review]
I agree with the renovate config part but afaik you've created a ticket already 👍 .
Looks good from my side, tests are green! In case of failure a bit more detailed is printed to actually debug it.
Renovate we had a look earlier and it captured all the versions 🚀.

For now I think it's fine to capture all renovate updates in a single PR. If we notice it makes sense to split, we can still do so later.

**Review by maxdanilov — COMMENTED** [review]
LGTM, thank you! Just one last comment about the `.tool-versions` sorting, but can be merged after that.

File: .tool-versions
Comment (maxdanilov) [reply .tool-versions:16]: tbf, I think that `.pre-commit-hooks/sort_tool_versions.py` is an overkill here. I don't see that list of tools changed often, and if it does, it should be a convention for the author to keep it sorted.

In case you want to have automation, I then would suggest a check that verifies if a list is sorted or not (and then the fix is up to the author). Something like

Comment (maxdanilov) [reply .tool-versions:16]: > I can use the proposed bash implementation, I will try it!

I'd say let's not spend too much time on it, if it's a quick win sure, otherwise as I mentioned totally fine to have PR authors check the sorting when introducing a new tool.

**Review by maxdanilov — APPROVED** [review]
LGTM, thank you! 🚀

**Pull request merged**
